src/sage/rings/number_field/number_field.py

```diff
@@ -156,24 +152,32 @@ def is_NumberFieldHomsetCodomain(codomain):
         sage: is_NumberFieldHomsetCodomain(NumberField(x^2 + 1, 'x'))
         True
         sage: is_NumberFieldHomsetCodomain(ZZ)
-        False
+        True
         sage: is_NumberFieldHomsetCodomain(3)
         False
         sage: is_NumberFieldHomsetCodomain(MatrixSpace(QQ, 2))
-        False
+        True
         sage: is_NumberFieldHomsetCodomain(InfinityRing)
-        False
-
-    Question: should, for example, QQ-algebras be accepted as well?
+        True
 
-    Caveat: Gap objects are not (yet) in :class:`Fields`, and therefore
-    not accepted as number field homset codomains::
+    Gap objects are not (yet) in :class:`Fields`, and therefore not accepted as
+    number field homset codomains::
 
         sage: is_NumberFieldHomsetCodomain(gap.Rationals)                               # needs sage.libs.gap
         False
     """
-    from sage.categories.fields import Fields
-    return codomain in Fields()
+    from sage.categories.rings import Rings
+
+    if category is None:
+        category = Rings()
+
+    if not category.is_subcategory(Rings()):
+        return False
+
+    if codomain not in category:
+        return False
```

Comment:
```suggestion
    assert codomain in category
```
I believe we agreed on this.

Reply:
I tried this, but then I get
```
sage -t --warn-long 5.0 --random-seed=119788034856104425102958228681385469939 src/sage/arith/misc.py
**********************************************************************
File "src/sage/arith/misc.py", line 2060, in sage.arith.misc.xgcd
Failed example:
    xgcd(y^2, a*h*y + b)
Exception raised:
    Traceback (most recent call last):
      File "sage/misc/cachefunc.pyx", line 1970, in sage.misc.cachefunc.CachedMethodCaller.__call__
        return cache[k]
    KeyError: (Number Field in g with defining polynomial x^2 - 3, None, True)

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/home/martin/sage/src/sage/doctest/forker.py", line 728, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/martin/sage/src/sage/doctest/forker.py", line 1152, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.arith.misc.xgcd[34]>", line 1, in <module>
        xgcd(y**Integer(2), a*h*y + b)
      File "/home/martin/sage/src/sage/arith/misc.py", line 2079, in xgcd
        return a.xgcd(b)
               ^^^^^^^^^
      File "sage/structure/element.pyx", line 4743, in sage.structure.element.coerce_binop.new_method
        return method(self, other, *args, **kwargs)
      File "sage/rings/polynomial/polynomial_element.pyx", line 9709, in sage.rings.polynomial.polynomial_element.Polynomial.xgcd
        return self.base_ring()._xgcd_univariate_polynomial(self, other)
      File "/home/martin/sage/src/sage/categories/fields.py", line 466, in _xgcd_univariate_polynomial
        q, r = d.quo_rem(v3)
               ^^^^^^^^^^^^^
      File "sage/structure/element.pyx", line 4743, in sage.structure.element.coerce_binop.new_method
        return method(self, other, *args, **kwargs)
      File "sage/rings/polynomial/polynomial_element.pyx", line 12538, in sage.rings.polynomial.polynomial_element.Polynomial_generic_dense.quo_rem
        q = x[n+k-1] * inv
      File "sage/structure/element.pyx", line 1505, in sage.structure.element.Element.__mul__
        return (<Element>left)._mul_(right)
      File "sage/rings/fraction_field_element.pyx", line 705, in sage.rings.fraction_field_element.FractionFieldElement._mul_
        d1 = rnum.gcd(sden)
      File "sage/rings/polynomial/multi_polynomial.pyx", line 2222, in sage.rings.polynomial.multi_polynomial.MPolynomial.gcd
        return self.parent()(doit(uniself, other.polynomial(x)))
      File "/home/martin/sage/src/sage/categories/unique_factorization_domains.py", line 220, in _gcd_univariate_polynomial
        _, R = A.pseudo_quo_rem(B)
               ^^^^^^^^^^^^^^^^^^^
      File "sage/rings/polynomial/polynomial_element.pyx", line 5409, in sage.rings.polynomial.polynomial_element.Polynomial.pseudo_quo_rem
        if other in self._parent.base_ring():
      File "sage/structure/parent.pyx", line 1170, in sage.structure.parent.Parent.__contains__
        x2 = self(x)
      File "sage/structure/parent.pyx", line 901, in sage.structure.parent.Parent.__call__
        return mor._call_(x)
      File "sage/rings/polynomial/polynomial_element.pyx", line 12926, in sage.rings.polynomial.polynomial_element.ConstantPolynomialSection._call_
        raise TypeError(f"{x} is not a constant polynomial")
      File "sage/structure/sage_object.pyx", line 221, in sage.structure.sage_object.SageObject.__repr__
        return reprfunc()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3190, in sage.rings.polynomial.polynomial_element.Polynomial._repr_
        return self._repr()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3154, in sage.rings.polynomial.polynomial_element.Polynomial._repr
        x = y = repr(x)
      File "sage/structure/sage_object.pyx", line 221, in sage.structure.sage_object.SageObject.__repr__
        return reprfunc()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3190, in sage.rings.polynomial.polynomial_element.Polynomial._repr_
        return self._repr()
      File "sage/rings/polynomial/polynomial_element.pyx", line 3154, in sage.rings.polynomial.polynomial_element.Polynomial._repr
        x = y = repr(x)
      File "sage/structure/sage_object.pyx", line 221, in sage.structure.sage_object.SageObject.__repr__
        return reprfunc()
      File "sage/rings/number_field/number_field_element.pyx", line 5052, in sage.rings.number_field.number_field_element.NumberFieldElement_relative._repr_
        return repr(R(self.list()))
      File "sage/rings/number_field/number_field_element.pyx", line 5019, in sage.rings.number_field.number_field_element.NumberFieldElement_relative.list
        return self.vector().list()
      File "sage/rings/number_field/number_field_element.pyx", line 3675, in sage.rings.number_field.number_field_element.NumberFieldElement.vector
        return self.number_field().relative_vector_space()[2](self)
      File "/home/martin/sage/src/sage/rings/number_field/number_field_rel.py", line 1456, in relative_vector_space
        return self.free_module(self.base_field(), *args, **kwds)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/misc/cachefunc.pyx", line 1975, in sage.misc.cachefunc.CachedMethodCaller.__call__
        w = self._instance_call(*args, **kwds)
      File "sage/misc/cachefunc.pyx", line 1851, in sage.misc.cachefunc.CachedMethodCaller._instance_call
        return self.f(self._instance, *args, **kwds)
      File "/home/martin/sage/src/sage/rings/number_field/number_field_rel.py", line 1413, in free_module
        to = maps.MapRelativeNumberFieldToRelativeVectorSpace(self, V)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/rings/number_field/maps.py", line 346, in __init__
        NumberFieldIsomorphism.__init__(self, Hom(K, V))
                                              ^^^^^^^^^
      File "/home/martin/sage/src/sage/categories/homset.py", line 418, in Hom
        H = Hom(X, Y, category, check=False)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/categories/homset.py", line 445, in Hom
        H = X._Hom_(Y, category)
            ^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/rings/number_field/number_field_rel.py", line 778, in _Hom_
        if not is_NumberFieldHomsetCodomain(codomain):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/martin/sage/src/sage/rings/number_field/number_field.py", line 177, in is_NumberFieldHomsetCodomain
        assert codomain in category, f"codomain (={codomain}) should be in the category (={category})"
               ^^^^^^^^^^^^^^^^^^^^
    AssertionError: codomain (=Vector space of dimension 2 over Number Field in g with defining polynomial x^2 - 3) should be in the category (=Category of rings)
```

Reply:
(and many similar failures in `schemes`)

Reply:
I think we need to be more careful about the default category. Rings is a bad choice as it doesn't take into account the inputs. I think it should be the meet of the (co)domain's categories.

Reply:
@tscrim : could you please turn your wish into a more precise suggestion ?

Reply:
I added it above (since only the codomain is given to the method here).

Reply:
My change above is also with making this an assert statement.

Reply:
I still would like this changed to an `assert` statement:
```suggestion
    assert codomain in category
```